src/lib.rs

```diff
@@ -1,31 +1,59 @@
-extern crate regex;
-extern crate crypto;
-
-use std::i64;
-use regex::Regex;
-use self::crypto::digest::Digest;
-use self::crypto::sha3::Sha3;
+#![cfg_attr(any(test, bench), feature(test))]
+use crypto::{digest::Digest, sha3::Sha3};
 
 pub fn checksum(address: &str) -> String {
-    let re = Regex::new(r"^0x").unwrap();
-    let address = address.to_lowercase();
-    let address = re.replace_all(&address, "").to_string();
-
-    let mut checksum_address = "0x".to_string();
-    let mut hasher = Sha3::keccak256();
-    hasher.input_str(&address);
-    let address_hash = hasher.result_str();
-
-    for i in 0..address.len() {
-        let n = i64::from_str_radix(&address_hash.chars().nth(i).unwrap().to_string(), 16).unwrap();
-        let ch = address.chars().nth(i).unwrap();
-        // make char uppercase if ith character is 9..f
-        if n > 7 {
-          checksum_address = format!("{}{}", checksum_address, ch.to_uppercase().to_string());
-        } else {
-          checksum_address = format!("{}{}", checksum_address, ch.to_string());
-        }
+    let address = address.trim_start_matches("0x").to_lowercase();
+
+    let address_hash = {
+        let mut hasher = Sha3::keccak256();
+        hasher.input(address.as_bytes());
+        hasher.result_str()
+    };
+
+    address
+        .char_indices()
+        .fold(String::from("0x"), |mut acc, (index, address_char)| {
+            // this cannot fail since it's Keccak256 hashed
+            let n = u16::from_str_radix(&address_hash[index..index + 1], 16).unwrap();
+
+            if n > 7 {
+                // make char uppercase if ith character is 9..f
```

Comment:
May as well fix the comment while you're here:
```suggestion
                // make char uppercase if ith character is 8..f
```